// @devvit/protos/types/devvit/plugin/redditapi/postcollections/postcollections_svc.twirp-client.js
import { CollectionRequest, CollectionResponse, SubredditCollectionsRequest, SubredditCollectionsResponse, CreateCollectionRequest, DeleteCollectionRequest, AddPostToCollectionRequest, RemovePostInCollectionRequest, FollowCollectionRequest, ReorderCollectionRequest, UpdateCollectionTitleRequest, UpdateCollectionDescriptionRequest, UpdateCollectionDisplayLayoutRequest, } from './postcollections_msg.js';
import { Empty } from '../../../../google/protobuf/empty.js';
export class PostCollectionsClientJSON {
    constructor(rpc) {
        this.rpc = rpc;
        this.Collection.bind(this);
        this.SubredditCollections.bind(this);
        this.Create.bind(this);
        this.Delete.bind(this);
        this.AddPost.bind(this);
        this.RemovePost.bind(this);
        this.Follow.bind(this);
        this.Reorder.bind(this);
        this.UpdateTitle.bind(this);
        this.UpdateDescription.bind(this);
        this.UpdateDisplayLayout.bind(this);
    }
    Collection(request) {
        const data = CollectionRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Collection', 'application/json', data);
        return promise.then((data) => CollectionResponse.fromJSON(data));
    }
    SubredditCollections(request) {
        const data = SubredditCollectionsRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'SubredditCollections', 'application/json', data);
        return promise.then((data) => SubredditCollectionsResponse.fromJSON(data));
    }
    Create(request) {
        const data = CreateCollectionRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Create', 'application/json', data);
        return promise.then((data) => CollectionResponse.fromJSON(data));
    }
    Delete(request) {
        const data = DeleteCollectionRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Delete', 'application/json', data);
        return promise.then((data) => Empty.fromJSON(data));
    }
    AddPost(request) {
        const data = AddPostToCollectionRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'AddPost', 'application/json', data);
        return promise.then((data) => Empty.fromJSON(data));
    }
    RemovePost(request) {
        const data = RemovePostInCollectionRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'RemovePost', 'application/json', data);
        return promise.then((data) => Empty.fromJSON(data));
    }
    Follow(request) {
        const data = FollowCollectionRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Follow', 'application/json', data);
        return promise.then((data) => Empty.fromJSON(data));
    }
    Reorder(request) {
        const data = ReorderCollectionRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Reorder', 'application/json', data);
        return promise.then((data) => Empty.fromJSON(data));
    }
    UpdateTitle(request) {
        const data = UpdateCollectionTitleRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'UpdateTitle', 'application/json', data);
        return promise.then((data) => Empty.fromJSON(data));
    }
    UpdateDescription(request) {
        const data = UpdateCollectionDescriptionRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'UpdateDescription', 'application/json', data);
        return promise.then((data) => Empty.fromJSON(data));
    }
    UpdateDisplayLayout(request) {
        const data = UpdateCollectionDisplayLayoutRequest.toJSON(request);
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'UpdateDisplayLayout', 'application/json', data);
        return promise.then((data) => Empty.fromJSON(data));
    }
}
export class PostCollectionsClientProtobuf {
    constructor(rpc) {
        this.rpc = rpc;
        this.Collection.bind(this);
        this.SubredditCollections.bind(this);
        this.Create.bind(this);
        this.Delete.bind(this);
        this.AddPost.bind(this);
        this.RemovePost.bind(this);
        this.Follow.bind(this);
        this.Reorder.bind(this);
        this.UpdateTitle.bind(this);
        this.UpdateDescription.bind(this);
        this.UpdateDisplayLayout.bind(this);
    }
    Collection(request) {
        const data = CollectionRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Collection', 'application/protobuf', data);
        return promise.then((data) => CollectionResponse.decode(data));
    }
    SubredditCollections(request) {
        const data = SubredditCollectionsRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'SubredditCollections', 'application/protobuf', data);
        return promise.then((data) => SubredditCollectionsResponse.decode(data));
    }
    Create(request) {
        const data = CreateCollectionRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Create', 'application/protobuf', data);
        return promise.then((data) => CollectionResponse.decode(data));
    }
    Delete(request) {
        const data = DeleteCollectionRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Delete', 'application/protobuf', data);
        return promise.then((data) => Empty.decode(data));
    }
    AddPost(request) {
        const data = AddPostToCollectionRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'AddPost', 'application/protobuf', data);
        return promise.then((data) => Empty.decode(data));
    }
    RemovePost(request) {
        const data = RemovePostInCollectionRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'RemovePost', 'application/protobuf', data);
        return promise.then((data) => Empty.decode(data));
    }
    Follow(request) {
        const data = FollowCollectionRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Follow', 'application/protobuf', data);
        return promise.then((data) => Empty.decode(data));
    }
    Reorder(request) {
        const data = ReorderCollectionRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'Reorder', 'application/protobuf', data);
        return promise.then((data) => Empty.decode(data));
    }
    UpdateTitle(request) {
        const data = UpdateCollectionTitleRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'UpdateTitle', 'application/protobuf', data);
        return promise.then((data) => Empty.decode(data));
    }
    UpdateDescription(request) {
        const data = UpdateCollectionDescriptionRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'UpdateDescription', 'application/protobuf', data);
        return promise.then((data) => Empty.decode(data));
    }
    UpdateDisplayLayout(request) {
        const data = UpdateCollectionDisplayLayoutRequest.encode(request).finish();
        const promise = this.rpc.request('devvit.plugin.redditapi.postcollections.PostCollections', 'UpdateDisplayLayout', 'application/protobuf', data);
        return promise.then((data) => Empty.decode(data));
    }
}
