// @devvit/protos/types/devvit/plugin/redditapi/postcollections/postcollections_svc.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { CollectionRequest, CollectionResponse, SubredditCollectionsRequest, SubredditCollectionsResponse, CreateCollectionRequest, DeleteCollectionRequest, AddPostToCollectionRequest, RemovePostInCollectionRequest, FollowCollectionRequest, ReorderCollectionRequest, UpdateCollectionTitleRequest, UpdateCollectionDescriptionRequest, UpdateCollectionDisplayLayoutRequest, } from './postcollections_msg.js';
import { Empty } from '../../../../google/protobuf/empty.js';
export var PostCollectionsMethod;
(function (PostCollectionsMethod) {
    PostCollectionsMethod["Collection"] = "Collection";
    PostCollectionsMethod["SubredditCollections"] = "SubredditCollections";
    PostCollectionsMethod["Create"] = "Create";
    PostCollectionsMethod["Delete"] = "Delete";
    PostCollectionsMethod["AddPost"] = "AddPost";
    PostCollectionsMethod["RemovePost"] = "RemovePost";
    PostCollectionsMethod["Follow"] = "Follow";
    PostCollectionsMethod["Reorder"] = "Reorder";
    PostCollectionsMethod["UpdateTitle"] = "UpdateTitle";
    PostCollectionsMethod["UpdateDescription"] = "UpdateDescription";
    PostCollectionsMethod["UpdateDisplayLayout"] = "UpdateDisplayLayout";
})(PostCollectionsMethod || (PostCollectionsMethod = {}));
export const PostCollectionsMethodList = [
    PostCollectionsMethod.Collection,
    PostCollectionsMethod.SubredditCollections,
    PostCollectionsMethod.Create,
    PostCollectionsMethod.Delete,
    PostCollectionsMethod.AddPost,
    PostCollectionsMethod.RemovePost,
    PostCollectionsMethod.Follow,
    PostCollectionsMethod.Reorder,
    PostCollectionsMethod.UpdateTitle,
    PostCollectionsMethod.UpdateDescription,
    PostCollectionsMethod.UpdateDisplayLayout,
];
export function createPostCollectionsServer(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.redditapi.postcollections',
        serviceName: 'PostCollections',
        methodList: PostCollectionsMethodList,
        matchRoute: matchPostCollectionsRoute,
    });
}
function matchPostCollectionsRoute(method, events) {
    switch (method) {
        case 'Collection':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Collection' };
                await events.onMatch(ctx);
                return handlePostCollectionsCollectionRequest(ctx, service, data, interceptors);
            };
        case 'SubredditCollections':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'SubredditCollections' };
                await events.onMatch(ctx);
                return handlePostCollectionsSubredditCollectionsRequest(ctx, service, data, interceptors);
            };
        case 'Create':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Create' };
                await events.onMatch(ctx);
                return handlePostCollectionsCreateRequest(ctx, service, data, interceptors);
            };
        case 'Delete':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Delete' };
                await events.onMatch(ctx);
                return handlePostCollectionsDeleteRequest(ctx, service, data, interceptors);
            };
        case 'AddPost':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'AddPost' };
                await events.onMatch(ctx);
                return handlePostCollectionsAddPostRequest(ctx, service, data, interceptors);
            };
        case 'RemovePost':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'RemovePost' };
                await events.onMatch(ctx);
                return handlePostCollectionsRemovePostRequest(ctx, service, data, interceptors);
            };
        case 'Follow':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Follow' };
                await events.onMatch(ctx);
                return handlePostCollectionsFollowRequest(ctx, service, data, interceptors);
            };
        case 'Reorder':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'Reorder' };
                await events.onMatch(ctx);
                return handlePostCollectionsReorderRequest(ctx, service, data, interceptors);
            };
        case 'UpdateTitle':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'UpdateTitle' };
                await events.onMatch(ctx);
                return handlePostCollectionsUpdateTitleRequest(ctx, service, data, interceptors);
            };
        case 'UpdateDescription':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'UpdateDescription' };
                await events.onMatch(ctx);
                return handlePostCollectionsUpdateDescriptionRequest(ctx, service, data, interceptors);
            };
        case 'UpdateDisplayLayout':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'UpdateDisplayLayout' };
                await events.onMatch(ctx);
                return handlePostCollectionsUpdateDisplayLayoutRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsCollectionRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsCollectionJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsCollectionProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsSubredditCollectionsRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsSubredditCollectionsJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsSubredditCollectionsProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsCreateRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsCreateJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsCreateProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsDeleteRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsDeleteJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsDeleteProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsAddPostRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsAddPostJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsAddPostProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsRemovePostRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsRemovePostJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsRemovePostProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsFollowRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsFollowJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsFollowProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsReorderRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsReorderJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsReorderProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsUpdateTitleRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsUpdateTitleJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsUpdateTitleProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsUpdateDescriptionRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsUpdateDescriptionJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsUpdateDescriptionProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handlePostCollectionsUpdateDisplayLayoutRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handlePostCollectionsUpdateDisplayLayoutJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handlePostCollectionsUpdateDisplayLayoutProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handlePostCollectionsCollectionJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = CollectionRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Collection(ctx, inputReq);
        });
    }
    else {
        response = await service.Collection(ctx, request);
    }
    return JSON.stringify(CollectionResponse.toJSON(response));
}
async function handlePostCollectionsSubredditCollectionsJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = SubredditCollectionsRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.SubredditCollections(ctx, inputReq);
        });
    }
    else {
        response = await service.SubredditCollections(ctx, request);
    }
    return JSON.stringify(SubredditCollectionsResponse.toJSON(response));
}
async function handlePostCollectionsCreateJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = CreateCollectionRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Create(ctx, inputReq);
        });
    }
    else {
        response = await service.Create(ctx, request);
    }
    return JSON.stringify(CollectionResponse.toJSON(response));
}
async function handlePostCollectionsDeleteJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = DeleteCollectionRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Delete(ctx, inputReq);
        });
    }
    else {
        response = await service.Delete(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handlePostCollectionsAddPostJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = AddPostToCollectionRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.AddPost(ctx, inputReq);
        });
    }
    else {
        response = await service.AddPost(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handlePostCollectionsRemovePostJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = RemovePostInCollectionRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RemovePost(ctx, inputReq);
        });
    }
    else {
        response = await service.RemovePost(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handlePostCollectionsFollowJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = FollowCollectionRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Follow(ctx, inputReq);
        });
    }
    else {
        response = await service.Follow(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handlePostCollectionsReorderJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = ReorderCollectionRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Reorder(ctx, inputReq);
        });
    }
    else {
        response = await service.Reorder(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handlePostCollectionsUpdateTitleJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = UpdateCollectionTitleRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateTitle(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateTitle(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handlePostCollectionsUpdateDescriptionJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = UpdateCollectionDescriptionRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateDescription(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateDescription(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handlePostCollectionsUpdateDisplayLayoutJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = UpdateCollectionDisplayLayoutRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateDisplayLayout(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateDisplayLayout(ctx, request);
    }
    return JSON.stringify(Empty.toJSON(response));
}
async function handlePostCollectionsCollectionProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = CollectionRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Collection(ctx, inputReq);
        });
    }
    else {
        response = await service.Collection(ctx, request);
    }
    return Buffer.from(CollectionResponse.encode(response).finish());
}
async function handlePostCollectionsSubredditCollectionsProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = SubredditCollectionsRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.SubredditCollections(ctx, inputReq);
        });
    }
    else {
        response = await service.SubredditCollections(ctx, request);
    }
    return Buffer.from(SubredditCollectionsResponse.encode(response).finish());
}
async function handlePostCollectionsCreateProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = CreateCollectionRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Create(ctx, inputReq);
        });
    }
    else {
        response = await service.Create(ctx, request);
    }
    return Buffer.from(CollectionResponse.encode(response).finish());
}
async function handlePostCollectionsDeleteProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = DeleteCollectionRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Delete(ctx, inputReq);
        });
    }
    else {
        response = await service.Delete(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handlePostCollectionsAddPostProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = AddPostToCollectionRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.AddPost(ctx, inputReq);
        });
    }
    else {
        response = await service.AddPost(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handlePostCollectionsRemovePostProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = RemovePostInCollectionRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.RemovePost(ctx, inputReq);
        });
    }
    else {
        response = await service.RemovePost(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handlePostCollectionsFollowProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = FollowCollectionRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Follow(ctx, inputReq);
        });
    }
    else {
        response = await service.Follow(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handlePostCollectionsReorderProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = ReorderCollectionRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.Reorder(ctx, inputReq);
        });
    }
    else {
        response = await service.Reorder(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handlePostCollectionsUpdateTitleProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = UpdateCollectionTitleRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateTitle(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateTitle(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handlePostCollectionsUpdateDescriptionProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = UpdateCollectionDescriptionRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateDescription(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateDescription(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
async function handlePostCollectionsUpdateDisplayLayoutProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = UpdateCollectionDisplayLayoutRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.UpdateDisplayLayout(ctx, inputReq);
        });
    }
    else {
        response = await service.UpdateDisplayLayout(ctx, request);
    }
    return Buffer.from(Empty.encode(response).finish());
}
