// @devvit/protos/types/devvit/plugin/redditapiv2/redditapiv2.twirp.js
import { TwirpServer, TwirpError, TwirpErrorCode, TwirpContentType, chainInterceptors, } from 'twirp-ts';
import { SubredditCollectionsRequest, SubredditCollectionsResponse, } from '../redditapi/postcollections/postcollections_msg.js';
import { DeleteFlairTemplateRequest } from '../redditapi/flair/flair_msg.js';
import { JsonStatus } from '../redditapi/common/common_msg.js';
export var RedditAPIV2Method;
(function (RedditAPIV2Method) {
    RedditAPIV2Method["GetSubredditCollections"] = "GetSubredditCollections";
    RedditAPIV2Method["DeleteFlairTemplate"] = "DeleteFlairTemplate";
})(RedditAPIV2Method || (RedditAPIV2Method = {}));
export const RedditAPIV2MethodList = [
    RedditAPIV2Method.GetSubredditCollections,
    RedditAPIV2Method.DeleteFlairTemplate,
];
export function createRedditAPIV2Server(service) {
    return new TwirpServer({
        service,
        packageName: 'devvit.plugin.redditapiv2',
        serviceName: 'RedditAPIV2',
        methodList: RedditAPIV2MethodList,
        matchRoute: matchRedditAPIV2Route,
    });
}
function matchRedditAPIV2Route(method, events) {
    switch (method) {
        case 'GetSubredditCollections':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'GetSubredditCollections' };
                await events.onMatch(ctx);
                return handleRedditAPIV2GetSubredditCollectionsRequest(ctx, service, data, interceptors);
            };
        case 'DeleteFlairTemplate':
            return async (ctx, service, data, interceptors) => {
                ctx = { ...ctx, methodName: 'DeleteFlairTemplate' };
                await events.onMatch(ctx);
                return handleRedditAPIV2DeleteFlairTemplateRequest(ctx, service, data, interceptors);
            };
        default:
            events.onNotFound();
            const msg = `no handler found`;
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleRedditAPIV2GetSubredditCollectionsRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleRedditAPIV2GetSubredditCollectionsJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleRedditAPIV2GetSubredditCollectionsProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
function handleRedditAPIV2DeleteFlairTemplateRequest(ctx, service, data, interceptors) {
    switch (ctx.contentType) {
        case TwirpContentType.JSON:
            return handleRedditAPIV2DeleteFlairTemplateJSON(ctx, service, data, interceptors);
        case TwirpContentType.Protobuf:
            return handleRedditAPIV2DeleteFlairTemplateProtobuf(ctx, service, data, interceptors);
        default:
            const msg = 'unexpected Content-Type';
            throw new TwirpError(TwirpErrorCode.BadRoute, msg);
    }
}
async function handleRedditAPIV2GetSubredditCollectionsJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = SubredditCollectionsRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetSubredditCollections(ctx, inputReq);
        });
    }
    else {
        response = await service.GetSubredditCollections(ctx, request);
    }
    return JSON.stringify(SubredditCollectionsResponse.toJSON(response));
}
async function handleRedditAPIV2DeleteFlairTemplateJSON(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        const body = JSON.parse(data.toString() || '{}');
        request = DeleteFlairTemplateRequest.fromJSON(body);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the json request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.DeleteFlairTemplate(ctx, inputReq);
        });
    }
    else {
        response = await service.DeleteFlairTemplate(ctx, request);
    }
    return JSON.stringify(JsonStatus.toJSON(response));
}
async function handleRedditAPIV2GetSubredditCollectionsProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = SubredditCollectionsRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.GetSubredditCollections(ctx, inputReq);
        });
    }
    else {
        response = await service.GetSubredditCollections(ctx, request);
    }
    return Buffer.from(SubredditCollectionsResponse.encode(response).finish());
}
async function handleRedditAPIV2DeleteFlairTemplateProtobuf(ctx, service, data, interceptors) {
    let request;
    let response;
    try {
        request = DeleteFlairTemplateRequest.decode(data);
    }
    catch (e) {
        if (e instanceof Error) {
            const msg = 'the protobuf request could not be decoded';
            throw new TwirpError(TwirpErrorCode.Malformed, msg).withCause(e, true);
        }
    }
    if (interceptors && interceptors.length > 0) {
        const interceptor = chainInterceptors(...interceptors);
        response = await interceptor(ctx, request, (ctx, inputReq) => {
            return service.DeleteFlairTemplate(ctx, inputReq);
        });
    }
    else {
        response = await service.DeleteFlairTemplate(ctx, request);
    }
    return Buffer.from(JsonStatus.encode(response).finish());
}
